File: Sources/SomnioCore/Placement/FeetMask.swift
import Foundation

/// Axis-aligned integer rectangle in pixel space. Endpoints are `Int32` so a rect derived from
/// a sprite near a sector edge cannot trap the `x + width` add. Right/bottom edges are exclusive,
/// matching `CollisionMaskOverlap`'s point-test polarity.
public struct PixelRect: Sendable, Equatable {
    public var x: Int32
    public var y: Int32
    public var width: Int32
    public var height: Int32

    public init(x: Int32, y: Int32, width: Int32, height: Int32) {
        self.x = x
        self.y = y
        self.width = width
        self.height = height
    }

    public var maxX: Int32 {
        x + width
    }

    public var maxY: Int32 {
        y + height
    }

    /// Geometric center, used by the proximity gates that compare feet positions rather than
    /// sprite top-left origins (the original's `maskxy + maskgroesse / 2`).
    public var center: (x: Int32, y: Int32) {
        (x + width / 2, y + height / 2)
    }
}

/// Derives an entity's feet collision box from its position and **sprite cell size**. The
/// original collision uses a bottom-aligned, full-width sub-rect of the sprite rather than the
/// whole cell, so a character's head can overlap a wall while the feet do not. Mirrors the
/// original `maskgroesse(0) = groesse(0)`, `maskgroesse(1) = groesse(1) / 4 + 4`,
/// `maskxy(0) = xy(0)`, `maskxy(1) = xy(1) + groesse(1) - maskgroesse(1)`
/// (`Somnio_Server.txt` decoded lines 389-390, 770-771). Lives next to `CollisionMaskOverlap`
/// so the client predictor and the server's per-sector actor derive the identical box.
public enum FeetMask {
    public static func rect(forSpriteAt position: GridPoint, spriteSize: GridSize) -> PixelRect {
        let feetHeight = Int32(spriteSize.height) / 4 + 4
        let feetWidth = Int32(spriteSize.width)
        let originX = Int32(position.x)
        let originY = Int32(position.y) + Int32(spriteSize.height) - feetHeight
        return PixelRect(x: originX, y: originY, width: feetWidth, height: feetHeight)
    }

    /// Feet-box center for proximity gates (NPC dialog radius, monster aggro, client bump/portal),
    /// replacing the old sprite-top-left center so gates trigger at the feet, not the head.
    public static func center(forSpriteAt position: GridPoint, spriteSize: GridSize) -> (x: Int32, y: Int32) {
        rect(forSpriteAt: position, spriteSize: spriteSize).center
    }

    /// `true` when the feet box of a `spriteSize` sprite at `position` lies within `sector` bounds
    /// and overlaps neither a static collision mask nor any `blockers` feet box. The single gate the
    /// client predictor and the server's per-sector actor both call, so a predicted move matches
    /// exactly what the actor accepts.
    public static func isClear(at position: GridPoint, spriteSize: GridSize, sector: Sector, blockers: [PixelRect]) -> Bool {
        let feet = rect(forSpriteAt: position, spriteSize: spriteSize)
        guard feet.x >= 0, feet.y >= 0,
              feet.maxX <= sector.pixelWidth, feet.maxY <= sector.pixelHeight
        else { return false }
        if CollisionMaskOverlap.intersects(feet, sector.collisionMasks) { return false }
        for blocker in blockers where CollisionMaskOverlap.overlaps(feet, blocker) {
            return false
        }
        return true
    }

    /// Clamps `position` so a `spriteSize` sprite's feet box stays within `sector` bounds. The
    /// movement predictor clamps its step target with this so a step toward an edge lands the feet
    /// box flush against it instead of stopping up to one tick short. Bounds mirror `isClear`'s
    /// (feet width == sprite width, feet bottom == sprite bottom); the inner `max` keeps a sector
    /// narrower/shorter than the sprite from inverting the clamp range.
    public static func clamped(_ position: GridPoint, spriteSize: GridSize, sector: Sector) -> GridPoint {
        let feetHeight = Int32(spriteSize.height) / 4 + 4
        let maxX = Int32(sector.pixelWidth) - Int32(spriteSize.width)
        let minY = feetHeight - Int32(spriteSize.height)
        let maxY = Int32(sector.pixelHeight) - Int32(spriteSize.height)
        return GridPoint(
            x: Int16(clamping: min(max(Int32(position.x), 0), max(0, maxX))),
            y: Int16(clamping: min(max(Int32(position.y), minY), max(minY, maxY)))
        )
    }
}

File: Web/src/core/feetMask.ts
import { clampToInt16, maxX, maxY, rectCenter } from './geometry'
import type { GridPoint, GridSize, PixelRect } from './geometry'
import { intersects, overlaps } from './collisionMaskOverlap'
import type { Sector } from './sector'
import { sectorPixelHeight, sectorPixelWidth } from './sector'

/**
 * Mirror of `Sources/SomnioCore/Placement/FeetMask.swift`. Collision uses a bottom-aligned,
 * full-width sub-rect of the sprite rather than the whole cell, so a character's head can
 * overlap a wall while the feet do not. The client predictor and the server's per-sector actor
 * derive the identical box — a divergence here shows up as the server rejecting moves the
 * client already drew.
 */

/** Bottom-aligned feet-box height: `spriteHeight / 4 + 4`, with Swift's integer division. */
export function feetHeight(spriteSize: GridSize): number {
  return Math.trunc(spriteSize.height / 4) + 4
}

export function feetRect(position: GridPoint, spriteSize: GridSize): PixelRect {
  const height = feetHeight(spriteSize)
  return {
    x: position.x,
    y: position.y + spriteSize.height - height,
    width: spriteSize.width,
    height,
  }
}

/**
 * Feet-box centre for the proximity gates (NPC dialog radius, monster aggro, bump/portal
 * triggers) — the feet, not the sprite's top-left.
 */
export function feetCenter(position: GridPoint, spriteSize: GridSize): { x: number; y: number } {
  return rectCenter(feetRect(position, spriteSize))
}

/**
 * `true` when the feet box at `position` lies within sector bounds and overlaps neither a
 * static collision mask nor any blocker's feet box. This is the single gate both the predictor
 * and the server call.
 */
export function isFeetClear(
  position: GridPoint,
  spriteSize: GridSize,
  sector: Sector,
  blockers: readonly PixelRect[]
): boolean {
  const feet = feetRect(position, spriteSize)
  if (feet.x < 0 || feet.y < 0) return false
  if (maxX(feet) > sectorPixelWidth(sector) || maxY(feet) > sectorPixelHeight(sector)) return false
  if (intersects(feet, sector.collisionMasks)) return false
  for (const blocker of blockers) {
    if (overlaps(feet, blocker)) return false
  }
  return true
}

/**
 * Clamps `position` so the feet box stays within sector bounds. The predictor clamps its step
 * target with this so a move toward an edge lands flush against it instead of stopping up to
 * one tick short. The inner `max` keeps a sector narrower than the sprite from inverting the
 * clamp range.
 */
export function clampToSector(position: GridPoint, spriteSize: GridSize, sector: Sector): GridPoint {
  const height = feetHeight(spriteSize)
  const limitX = sectorPixelWidth(sector) - spriteSize.width
  const minY = height - spriteSize.height
  const limitY = sectorPixelHeight(sector) - spriteSize.height
  return {
    x: clampToInt16(Math.min(Math.max(position.x, 0), Math.max(0, limitX))),
    y: clampToInt16(Math.min(Math.max(position.y, minY), Math.max(minY, limitY))),
  }
}
